### ISSUE TITLE
Inconsistent behaviour when parameter appears multiple times in parameter list

### ISSUE THREAD

**Issue by pyscorcher** [pytorch/pytorch#40967]
## 🐛 Bug

When we pass a list of parameters or parameter groups to an optimizer, and one parameter appears multiple times we get different behaviours, and it is not clear whether this is intended that way:

* If the parameter appears twice within one parameter group, everything works. That parameter will get updated *twice* though.
* If the parameter appears in distinct parameter groups, then we get an error.

## To Reproduce

```
import torch
x = torch.zeros((1,), requires_grad=True)
# uncomment one of the following three lines for each of the cases:
# a = torch.optim.SGD(params=[dict(params=[x])], lr=0.1) # baseline
# a = torch.optim.SGD(params=[dict(params=[x, x])], lr=0.1)  # apparently acceptable (?); x is updated twice
# a = torch.optim.SGD(params=[dict(params=x), dict(params=x)], lr=0.1)  # apparently not acceptable
x.sum().backward()
a.step()
print(x)
```

## Expected behavior

I would expect that no matter in what way a parameter apperas multiple times, we get the same behaviour in both situations: Either we get errors in both situations, or both situations are deemed acceptable.

## Environment

 - PyTorch Version (e.g., 1.0): 1.5
 - OS (e.g., Linux): Win/Linux
 - How you installed PyTorch: conda
 - Python version: 3.7

**Comment by mariosasko** [comment]
I believe this is a valid behavior and there is no reason to change it. The idea of parameter groups is to enable the user to have a different set of optimizer parameters (lr, weight decay, ...) for each group. If the last definition was allowed, it would mean that it's possible to have ambiguous param groups that share same parameters with different optimizer policies, e.g. `torch.optim.SGD([{'params': x, 'lr': 0.01'}, {'params': x, 'lr': 0.05}]) # Which lr to choose for x?` 
@ezyang @apaszke @colesbury Whats your take on this issue? Maybe the param group with duplicate parameters could raise an error (even though I think it's fine as it is). If you have a better idea, I would like to implement it and make a PR.

**Comment by pyscorcher** [comment]
Thanks for your reply! It really is not a big problem, I just thought it would be nice to have that documented somewhere. Or alternatively I imagine one could implement a warning when the same parameter is present multiple times in the same parameter group.

**Comment by ezyang** [comment]
I'm not sure! Maybe triage review will know.

**Comment by ezyang** [comment]
@ngimel thinks most cases should be errors

**Comment by mariosasko** [comment]
I can take this one. Will make a PR soon (with a test).

**Comment by mariosasko** [comment]
@ezyang @albanD Made a PR and it passes all the checks. Now both cases raise errors.

**Comment by realiti4** [comment]
Hi,

I have been trying to use one optimizer for both actor and critic with this:

`torch.optim.Adam(list(actor.parameters()) + list(critic.parameters()), lr=lr)`

But I am getting
```
UserWarning: optimizer contains a parameter group with duplicate parameters; in future, this will cause an error; see github.com/pytorch/pytorch/issues/40967 for more information
```

What is the correct way of what I am trying to do now? Thanks.

**Comment by JermainLiu** [comment]
The idea of parameter groups is to enable the user to have a different set of optimizer parameters (lr, weight decay, ...) for each group. We can use `torch.optim.SGD([{'params': x, 'lr': 0.001'}, {'params': y, 'lr': 0.01}], lr=0.2)` to optimize `x` with `lr=0.01` and optimize `y` with `lr=0.1` .

**Comment by externalsupplierstaff** [comment]
> I have been trying to use one optimizer for both actor and critic with this:
> 
> `torch.optim.Adam(list(actor.parameters()) + list(critic.parameters()), lr=lr)`
Same question here:
> What is the correct way of what I am trying to do now? Thanks.

Tianshou RL library is well known and based in pytorch. However, in the tests that line is suggested.

How do we do union?

**Comment by externalsupplierstaff** [comment]
Just do `set(list(actor.parameters()) + list(critic.parameters()), lr=lr)`

**Comment by LeeShiyang** [comment]
Hi,

I also met similar issues. Here is what my model looks like

```
class GPT2PrefixLMHeadModel(GPT2PreTrainedModel):
    _keys_to_ignore_on_load_missing = [r"attn.masked_bias", r"attn.bias", r"lm_head.weight"]
    def __init__(self, config):
        super().__init__(config)
        self.transformer = GPT2Model(config)
        self.lm_head = nn.Linear(config.n_embd, config.vocab_size, bias=False)
        self.dropout = torch.nn.Dropout(config.prefix_dropout_prob)
        self.init_weights()
        self.model_parallel = False
        self.device_map = None
        self.pre_seq_len = config.pre_seq_len
        self.n_layer = config.num_hidden_layers
        self.n_head = config.num_attention_heads
        self.n_embd = config.hidden_size // config.num_attention_heads
        self.prefix_tokens = torch.arange(self.pre_seq_len).long()
        self.prefix_encoder = PrefixEncoder(config)
```
I create two different optimizers for the model. 

```
    def concat_generators(*args):
      for gen in args:
          yield from gen

    gpt2_params = concat_generators(model.transformer.parameters(),model.lm_head.parameters())
    optimizer_gpt2 = AdamW(gpt2_params, lr=args.learning_rate)
    optimizer_prefix = AdamW(model.prefix_encoder.parameters(), lr=args.learning_rate)
```
The above code will cause the following warnings:
`/home/shiyangli/anaconda3/envs/naacl/lib/python3.6/site-packages/transformers/optimization.py:311: UserWarning: optimizer contains a parameter group with duplicate parameters; in future, this will cause an error; see github.com/pytorch/pytorch/issues/40967 for more information`

This is not expected since model.transformer.parameters() and model.lm_head.parameters() don't have overlapping parameters. Why will this still happen? Your clarification is much appreciated.

**Comment by imsaumil** [comment]
Any updates on this ?

I have two identical models with different training datasets within the same training loop.

Is it okay to use `optimizer = torch.optim.Adam(list(vae_models[0].parameters()) + list(vae_models[1].parameters()), lr=args.learning_rate)` ?

**Comment by magicse** [comment]
I get warning about - parameter appears multiple times in parameter list
```
# Optimizers
optimizer_G = torch.optim.Adam(
    itertools.chain(encoder.parameters(), G1.parameters(), G2.parameters()),
    lr=opt.lr,
    betas=(opt.b1, opt.b2),
)

```
If i'll change to this . Is it correct ?

```
unique_params = set()
for name, param in itertools.chain(encoder.named_parameters(), G1.named_parameters(), G2.named_parameters()):
    if param.requires_grad:
        unique_params.add(param)

optimizer_G = torch.optim.Adam(unique_params, lr=opt.lr, betas=(opt.b1, opt.b2))
```
